src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,672 @@
+
+	if(state != m_clock_state) {
+		m_clock_state = state;
+		if(!state)
+			transmit_edge();
+		else
+			receive_edge();
+	}
+}
+
+void z180asci_channel_base::transmit_edge()
+{
+	m_tx_counter++;
+	if (m_tx_counter != m_divisor) return;
+	m_tx_counter = 0;
+
+	if (m_asci_cntla & Z180_CNTLA_TE)
+	{
+		switch (m_tx_state)
+		{
+		case STATE_START:
+			output_txa(0);
+			m_tx_state = STATE_DATA;
+			m_tx_parity = 0;
+			break;
+		case STATE_DATA:
+			m_tx_bits++;
+			output_txa(BIT(m_tsr, 0));
+			m_tx_parity ^= BIT(m_tsr, 0);
+			m_tsr >>= 1;
+			if (m_tx_bits == ((m_asci_cntla & Z180_CNTLA_MODE_DATA) ? 8 : 7))
+			{
+				m_tx_state = (m_asci_cntlb & Z180_CNTLB_MP) ? STATE_MPB : (m_asci_cntla & Z180_CNTLA_MODE_PARITY) ? STATE_PARITY : STATE_STOP;
+				m_tx_bits = (m_asci_cntla & Z180_CNTLA_MODE_STOPB) ? 2 : 1;
+			}
+			break;
+		case Z180_CNTLB_MP:
+			m_tx_state = STATE_STOP;
+			output_txa(m_asci_cntlb & Z180_CNTLB_MPBT ? 1 : 0);
+			break;
+		case STATE_PARITY:
+			m_tx_state = STATE_STOP;
+			if (m_asci_cntlb & Z180_CNTLB_PEO) m_tx_parity ^= 1; // odd parity
+			output_txa(m_tx_parity);
+			break;
+		case STATE_STOP:
+			if (m_tx_bits)
+			{
+				m_tx_bits--;
+				if (m_tx_bits == 0)
+				{
+					output_txa(1);
+					m_asci_stat |= Z180_STAT_TDRE;
+					if (m_asci_stat & Z180_STAT_TIE)
+					{
+						m_irq = 1;
+					}
+				}
+			}
+			break;
+		case STATE_BREAK:
+			output_txa(0);
+			break;
+		}
+	}
+}
+
+void z180asci_channel_base::receive_edge()
+{
+	m_rx_counter++;
+	if (m_rx_counter != m_divisor) return;
+	m_rx_counter = 0;
+
+	if (!m_rx_enabled) return;
+
+	if (m_asci_cntla & Z180_CNTLA_RE)
+	{
+		switch (m_rx_state)
+		{
+		case STATE_START:
+			if(m_rxa == 0)
+			{
+				m_rx_state = STATE_DATA;
+				m_rx_bits = 0;
+				m_rx_parity = 0;
+				m_rx_error = 0;
+				m_rsr = 0;
+			}
+			break;
+		case STATE_DATA:
+			m_rsr |= m_rxa << m_rx_bits;
+			m_rx_parity ^= m_rxa;
+			m_rx_bits++;
+			if (m_rx_bits == ((m_asci_cntla & Z180_CNTLA_MODE_DATA) ? 8 : 7))
+			{
+				m_rx_bits = (m_asci_cntla & Z180_CNTLA_MODE_STOPB) ? 2 : 1;
+				m_rx_state = (m_asci_cntlb & Z180_CNTLB_MP) ? STATE_MPB : (m_asci_cntla & Z180_CNTLA_MODE_PARITY) ? STATE_PARITY : STATE_STOP;
+			}
+			break;
+		case STATE_MPB:
+			m_rx_state = STATE_STOP;
+			m_asci_cntla |= m_rxa ? Z180_CNTLA_MPBR_EFR : 0;
+			break;
+		case STATE_PARITY:
+			m_rx_state = STATE_STOP;
+			if (m_asci_cntlb & Z180_CNTLB_PEO) m_rx_parity ^= 1; // odd parity
+			if (m_rx_parity != m_rxa)
+				m_rx_error |= Z180_STAT_PE;
+			break;
+		case STATE_STOP:
+			if (m_rxa == 0)
+				m_rx_error |= Z180_STAT_FE;
+
+			m_rx_bits--;
+			if (m_rx_bits == 0)
+			{
+				if ((m_rsr == 0) && (m_rx_parity == 0) && (m_rxa == 0))
+					m_asci_ext |= Z180_ASEXT_BRK_DET;
+
+				// Skip only if MPE mode active and MPB is 0
+				if (!((m_asci_cntla & Z180_CNTLA_MPE) && ((m_asci_cntla & Z180_CNTLA_MPBR_EFR) == 0)))
+				{
+					set_fifo_data(m_rsr, m_rx_error);
+				}
+				m_rx_state = STATE_START;
+			}
+			break;
```

Comment:
Break detection isn’t going to work properly either if you’re only sampling on bit intervals.  In reality, it will detect a space condition on any edge of the BRG output.